core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,281 @@
+
+  private fun getFeatures(tts: TextToSpeech?): Set<String> = tts?.voice?.features ?: setOf()
+
+  private fun loadURL(webView: WebView) {
+    // We use JavaScript to get the content of the page conveniently, earlier making some
+    // changes in the page
+    webView.loadUrl(
+      """
+        javascript:
+        body = document.getElementsByTagName('body')[0].cloneNode(true);
+        toRemove = body.querySelectorAll('sup.reference, #toc, .thumbcaption,     title, .navbox');
+        Array.prototype.forEach.call(toRemove, function(elem) {    
+          elem.parentElement.removeChild(elem);});
+        tts.speakAloud(body.innerText);
+        """.trimIndent()
+    )
+  }
+
+  fun stop() {
+    if (tts.stop() == SUCCESS) {
+      currentTTSTask = null
+      tts.setOnUtteranceProgressListener(null)
+      onSpeakingListener.onSpeakingEnded()
+    }
+  }
+
+  private fun requestAudioFocus(): Boolean {
+    val audioFocusRequest = am.requestAudioFocus(
+      onAudioFocusChangeListener, AudioManager.STREAM_MUSIC,
+      AudioManager.AUDIOFOCUS_GAIN
+    )
+    Log.d(TAG_KIWIX, "Audio Focus Requested")
+    synchronized(focusLock) {
+      return@requestAudioFocus audioFocusRequest ==
+        AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    }
+  }
+
+  fun pauseOrResume() {
+    currentTTSTask?.let {
+      if (it.paused) {
+        requestAudioFocus()
+      } else {
+        it.pause()
+      }
+    }
+  }
+
+  fun initWebView(webView: WebView) {
+    webView.addJavascriptInterface(TTSJavaScriptInterface(), "tts")
+  }
+
+  /**
+   * Releases the resources used by the engine.
+   *
+   * @see android.speech.tts.TextToSpeech.shutdown
+   */
+  fun shutdown() {
+    tts.shutdown()
+  }
+
+  /**
+   * The listener which is notified when initialization of the TextToSpeech engine is successfully
+   * done.
+   */
+  internal interface OnInitSucceedListener {
+    fun onInitSucceed()
+  }
+
+  /**
+   * The listener that is notified when speaking starts or stops (regardless of whether it was a
+   * result of error, user, or because whole text was read).
+   *
+   *
+   * Note that the methods of this interface may not be called from the UI thread.
+   */
+  interface OnSpeakingListener {
+    fun onSpeakingStarted()
+    fun onSpeakingEnded()
+  }
+
+  inner class TTSTask(private val pieces: List<String>) {
+    private val currentPiece =
+      AtomicInteger(0)
+    @JvmField var paused = true
+    fun pause() {
+      paused = true
+      currentPiece.decrementAndGet()
+      tts.setOnUtteranceProgressListener(null)
+      tts.stop()
+    }
+
+    fun start() {
+      if (!paused) {
+        return
+      }
+      paused = false
+      val params =
+        hashMapOf<String, String>()
+      // The utterance ID isn't actually used anywhere, the param is passed only to force
+      // the utterance listener to be notified
+      params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = "kiwixLastMessage"
```

Comment:
we were declaring this inside the `hashMapOf` nee `mapOf`, move it back there

Reply:
```
    val params =
        mutableMapOf<String, String>()

      params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = "kiwixLastMessage"
```
@macgills is this ok?

Reply:
`hashMapOf(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID to "kiwixLastMessage")`